### ISSUE TITLE
Invalid plan error with simple .tf file

### ISSUE THREAD

**Issue by SpencerBrown** [hashicorp/terraform-provider-tls#43]
### Terraform Version
Latest master of both terraform and provider as of Feb 20 2018

### Affected Resource(s)
tls_private_key

### Terraform Configuration Files
```hcl
resource tls_private_key "test" {
  algorithm = "RSA"
}
```

### Output

```
Spencer-Brown:t12 spencer$ terraform apply

Error: Provider produced invalid plan

Provider "tls" planned an invalid value for tls_private_key.test.rsa_bits:
planned value cty.NumberIntVal(2048) does not match config value
cty.NullVal(cty.Number).

This is a bug in the provider, which should be reported in the provider's own
issue tracker.


Error: Provider produced invalid plan

Provider "tls" planned an invalid value for tls_private_key.test.ecdsa_curve:
planned value cty.StringVal("P224") does not match config value
cty.NullVal(cty.String).

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

**Comment by SpencerBrown** [comment]
Note that this also fails:

```hcl
resource tls_private_key "test" {
  algorithm = "RSA"
  rsa_bits = 2048
}
```

with the following error:

```
Spencer-Brown:t12 spencer$ terraform apply

Error: Provider produced invalid plan

Provider "tls" planned an invalid value for tls_private_key.test.ecdsa_curve:
planned value cty.StringVal("P224") does not match config value
cty.NullVal(cty.String).

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

but this works:

```hcl
resource tls_private_key "test" {
  algorithm = "RSA"
  rsa_bits = 2048
  ecdsa_curve = ""
}
```

Something has changed in the core very recently to cause this issue in the provider.  A mismatch in types between Null Value from the config vs. a default value in the provider?

**Comment by apparentlymart** [comment]
Hi @SpencerBrown! Thanks for reporting this.

Terraform Core is indeed now being more careful to verify correct behavior of providers; Terraform used to trust providers to behave consistently but as we get ready to make third-party plugin development more accessible in a forthcoming release (auto-install behavior, etc) we've introduced some additional safety checks to ensure that providers behave as expected, which has in turn surfaced some bugs in existing providers that we need to address.

In this particular case it looks like the provider is populating some default values into the result that are not relevant to the given configuration -- in this case, the `ecdsa_curve` argument when generating an RSA key. However, I believe this particular check should be opted-out by the SDK as legacy behavior, so we may have to investigate why the check was still applied. I have a feeling it's because this provider currently has a slightly-older version of the SDK vendored into it which needs to be updated before release to include that opt-out.

We produced some dev snapshot builds on Friday that include provider builds with the SDK forced to latest (ignoring what was in `vendor/`), so I think the easiest way to verify this would be to try [that snapshot build of the `tls` provider](http://terraform-0.12.0-dev-snapshots.s3-website-us-west-2.amazonaws.com/terraform-provider-tls/2.0.0-dev20190216H00-dev/), which should include the opt-out I mentioned so that this older provider is permitted to make this unexpected change, which will then be resolved properly (making this attribute not be set at all) in a later release. If the snapshot build works, then the fix here will be to update the vendored SDK version to latest, which we'll do as part of the v0.12.0 beta1 release process.
